### DISCUSSION TITLE
WaitUntilDisplayed - default ImplicitMaxWait

### DISCUSSION THREAD

**Discussion by dornech** [GCuser99/SeleniumVBA#123] [General]
Just a quick question: for a consistent user/programmer experience - wouldn't it be logical to use the ImplicitMaxWait as default for the max wait time in WaitUntilDisplayed ?

**Comment by GCuser99** [comment]
We designed `WaitUntilDisplayed` based on the knowledge that `ImplicitMaxWait` (`ImplicitWait` in other bindings) only applies to element "presence" not "displayed-ness". Also, the Selenium Webdriver default for implicit wait is 0. While I can definitely see the logic of what you are suggesting, I think the way we designed it by prioritizing compatibility with Selenium, makes sense too. But of course, the "logic" of it depends on your priorities and expectations.

**Comment by Arvin-fang-ui** [comment]
about "WaitUntilDisplayed", could you teach me how can i write it?  set xx = driver.findelementxpat("xxxxxxx").WaitUntilDisplayed(1000), is this format right?

**Comment by 6DiegoDiego9** [comment]
Yes @Arvin-fang-ui, both these formats are fine:
set xx = driver.FindElementsByXPath("xxxxxxx").WaitUntilDisplayed(1000)  'returning the element, for future use
driver.FindElementsByXPath("xxxxxxx").WaitUntilDisplayed 1000  'if you will not need to use the element

**Reply by Arvin-fang-ui** [reply]
but if I  want  all the  web page elements are displayed and then click  this  element ，how can I  do  it ？appreciate  your  reply.

**Reply by hanamichi77777** [reply]
Hello @Arvin-fang-ui
I misunderstood how to use "Displayed" at first.
Maybe you want to wait until the element is displayed when the page loads?
If so, you can't use "Displayed".
You need a different solution.
For example, if you set the implicit wait to 5 seconds, FindElement will wait up to 5 seconds until the element is found before clicking it.

driver.ImplicitMaxWait = 5000
driver.FindElement(By.XPath,"xxxxxxx").click

**Reply by Arvin-fang-ui** [reply]
hi @hanamichi77777, thank you for the detailed reply. I usd "do loop" function to realize it. and for now, I meet another problem. Why is there no "driver.closewindow" function? How can I use handles to control the new tab?

**Comment by GCuser99** [comment]
https://github.com/GCuser99/SeleniumVBA/wiki#managing-windows

**Reply by Arvin-fang-ui** [reply]
Hi Mike @GCuser99, thank you for your reply, the information helps me a lot. The code seems good and runs with no problem. when I set the default download older as the pic shows. There are two questions. 1, if one line code is not written in, there is a warning like the pic shows. 2. if add this line, all files can be downloaded. but it will create one new folder named "False" in the destination, as the following pic shows. Could you give me some suggestions? have a nice day!
![SS](https://github.com/user-attachments/assets/b99a1723-b050-444c-964a-d71fc30a7412)
![sss](https://github.com/user-attachments/assets/39a266a2-84d8-42e1-9c8e-314f607f3e22)

**Comment by GCuser99** [comment]
It's hard for me to tell because I don't know what line in your code the automation error occurs. Probably on a line that is not shown in your pic?

It's much easier for us to try to help you if you can give us minimal reproducible code that we can try on our own systems to help debug. So I'll help get this started - try this and let us know what happens:

```vba
Sub test_file_download()
    Dim driver As SeleniumVBA.WebDriver
    Dim caps As SeleniumVBA.WebCapabilities
    
    Set driver = SeleniumVBA.New_WebDriver
    
    'assign your folder path here - you can use Environ("USERPROFILE") to hide your username(s)
    downloadFolderPath = Environ("USERPROFILE") & "\Desktop\New folder"
    
    driver.DefaultIOFolder = downloadFolderPath
    
    'qc default io path - these two should be same
    Debug.Assert driver.DefaultIOFolder = driver.ResolvePath("./")
    Debug.Print driver.DefaultIOFolder

    driver.StartEdge ("C:\SeleniumBasic\msedgedriver.exe")
    
    Set caps = driver.CreateCapabilities
    
    caps.SetDownloadPrefs downloadFolderPath:=downloadFolderPath
    
    driver.OpenBrowser caps

    driver.NavigateTo "https://www.selenium.dev/selenium/web/downloads/download.html"
    driver.Wait 500
    
    'start with a clean folder
    driver.DeleteFiles downloadFolderPath & "\file_1.txt", downloadFolderPath & "\file_2.jpg"
    
    'download the files
    driver.FindElementByCssSelector("#file-1").Click
    driver.WaitForDownload downloadFolderPath & "\file_1.txt"
    
    driver.FindElementByCssSelector("#file-2").Click
    driver.WaitForDownload downloadFolderPath & "\file_2.jpg"
    
    'if we get here then so far so good!
    'let's try again using paths relative to DefaultIOFolder
    
    driver.DeleteFiles ".\file_1.txt", ".\file_2.jpg"
    
    driver.FindElementByCssSelector("#file-1").Click
    driver.WaitForDownload ".\file_1.txt"
    
    driver.FindElementByCssSelector("#file-2").Click
    driver.WaitForDownload ".\file_2.jpg"
    
    driver.CloseBrowser
    driver.Shutdown
End Sub
```

**Reply by Arvin-fang-ui** [reply]
thank you, Mike. Your example gives me immense help.  the line mentioned in my question is one row code " driver.DefaultIOFolder = downloadFolderPath", now it seems that it's necessary. and for the second question, "caps.SetDownloadPrefs downloadfolderpath:=DownloadFolder" is correct, not"caps.SetDownloadPrefs downloadfolderpath =DownloadFolder".

**Comment by GCuser99** [comment]
The line `driver.DefaultIOFolder = downloadFolderPath` is only needed if you are using relative paths in your code such as `driver.DeleteFiles ".\file_1.txt", ".\file_2.jpg"`. If you use absolute paths, such as `driver.DeleteFiles downloadFolderPath & "\file_1.txt", downloadFolderPath & "\file_2.jpg"` then you should not need that line. If you do use relative paths and do not specify the `DefaultIOFolder`, then it defaults to the folder location of your Excel workbook.

One more thing - why are you specifying `"C:\SeleniumBasic\msedgedriver.exe"` as the driver location? Look in your DownLoads folder - the driver should also be there too. SeleniumVBA automatically downloads the latest driver version to that folder so that you do not need to do that manually. See [here](https://github.com/GCuser99/SeleniumVBA/wiki#getting-started) for more info.
